master/internal/api_experiment.go

```diff
@@ -673,16 +764,25 @@ func (a *apiServer) UnarchiveExperiment(
 func (a *apiServer) PatchExperiment(
 	ctx context.Context, req *apiv1.PatchExperimentRequest,
 ) (*apiv1.PatchExperimentResponse, error) {
-	var exp experimentv1.Experiment
-	switch err := a.m.db.QueryProto("get_experiment", &exp, req.Experiment.Id); {
-	case err == db.ErrNotFound:
-		return nil, status.Errorf(codes.NotFound, "experiment not found: %d", req.Experiment.Id)
-	case err != nil:
-		return nil, errors.Wrapf(err, "error fetching experiment from database: %d", req.Experiment.Id)
+	curUser, _, err := grpcutil.GetUser(ctx, a.m.db, &a.m.config.InternalConfig.ExternalSessions)
+	if err != nil {
+		return nil, err
+	}
+	exp, err := a.getExperiment(*curUser, int(req.Experiment.Id))
+	if err != nil {
+		return nil, err
+	}
+	modelExp, err := model.ExperimentFromProto(exp)
+	if err != nil {
+		return nil, err
 	}
 
 	madeChanges := false
 	if req.Experiment.Name != nil && exp.Name != req.Experiment.Name.Value {
+		if err = expauth.AuthZProvider.Get().CanSetExperimentsName(*curUser, modelExp); err != nil {
```

Comment:
I am worried this is way too granular. what's the case for allowing setting experiment name, but not the notes? 

maybe have `CanEditExperiment` for writes on general metadata and `CanEditExperimentArtifacts` for writes on "sensitive" stuff like the code or checkpoints. that'd be sufficient to map to our initial roles. we can break it down later. that'd also keep the authz API surface smaller.

on second thought, RBAC AuthZ implementation could do the same simplification internally (e.g. all of `CanSetExperimentsName`, `CanPauseExperiment` all calling internal `CanEditExperiment`).

Reply:
I agree it is way too excessive now, I'll merge a bunch of ```AuthZ``` methods together now. I think we could do the simplification internally but that seems like more work for no benefit. 

I do think for some cases like ```CanSetExperimentsWeight(curUser model.User, e *model.Experiment, weight float64)``` some people would be happy to see the feature to limit scheduling priority ranges to groups eventually

Reply:
Ok let me know what you think of the new smaller interface.

Reply:
looks good 👍 

my primary concern is the permissions in the world where we allow custom roles. if we go coarse today and need to add granularity later, we can do a SQL migration where coarse `get_experiment_artifacts` permission is split into `get_experiment_code` and `get_experiment_checkpoints`. this won't be breaking change.